pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,151 @@
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) error {
+	_, _, err := net.ParseCIDR(cidr)
+	return err
+}
+
+// GetUnreservedIPBlock returns an unreserved /29 IP block. It accepts the requested
+// CIDR and list of reserved IPs as arguments and returns the /29 IP available in that CIDR
+// Reserved IPV4 ranges can be available in the cloud instances or not as it is yet to be written.
+// All the Reserved IP ranges present in the cloud provider are passed to this function
+// All the Reserved IP ranges not present in the cloud provider are extracted from the reserved list in the IPAllocator
+// Finally a final reserved list is created by mergin these two lists
+// Potential error cases: 1) No /29 Block in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPBlock(cidr string, reservedIPRanges map[string]bool) (string, error) {
+	for reservedIPRange := range ipAllocator.reserved {
+		reservedIPRanges[reservedIPRange] = true
+	}
+
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("Unable to parse CIDR %s", cidr)
+	}
+	buffer := bytes.NewBufferString("")
+	for cidrIPBlock := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIPBlock) && err == nil; err = incrementIP(cidrIPBlock, incrementStep29IPBlock) {
+		overLap := false
+		buffer.WriteString(cidrIPBlock.String())
+		buffer.WriteString("/29")
+		ipRange := buffer.String()
+		for reservedIPRange := range reservedIPRanges {
+			// Find if the current IP range in the CIDR overlaps with any of the reserved IP ranges. If not, this can be returned
+			overLap = isOverlap(ipRange, reservedIPRange)
+			if overLap {
+				break
+			}
+		}
+		if !overLap {
+			return ipRange, nil
+		}
+		buffer.Reset()
+	}
+
+	// No unreserved IP range available in the entire CIDR range. Returning error
+	return "", fmt.Errorf("All of the /29 IP ranges in the cidr %s are reserved", cidr)
+}
+
+// validateCIDROverlap checks if two cidrs have any overlapping IPs
+func isOverlap(cidr1 string, cidr2 string) bool {
+	_, ipnet1, err := net.ParseCIDR(cidr1)
+	// Invalid CIDRs are considered as overlapping
+	if err != nil {
+		return true
+	}
+
+	_, ipnet2, err := net.ParseCIDR(cidr2)
+	if err != nil {
+		return true
+	}
+
+	return ipnet1.Contains(ipnet2.IP) || ipnet2.Contains(ipnet1.IP)
+}
+
+// Increment the given IP value by the provided step
+func incrementIP(ip net.IP, step byte) error {
+	ip = ip.To4()
+	for ipByte := 3; ipByte >= 0; ipByte-- {
+		//	If it the least significant byte, add step to it
+		if ipByte == 3 {
+			ip[ipByte] += step
+			// Addition did not result in a carry. We have the incremented value
+			if ip[ipByte] >= step {
+				break
+			}
+		} else {
+			// Add the carry from the previous addition
+			ip[ipByte]++
+
+			// Addition did not result in a carry. We have the incremented value
+			if ip[ipByte] > 0 {
```

Comment:
why 0?

Reply:
Because in case of most significant byte, the rollover will occur when your value changes from 255 to 0. Hence after adding the carry we check if the value is greater than 0 for MSB